File: Sources/PDFAnalyzerAppCore/Models/AnalysisResult.swift
import Foundation

// MARK: - Top-level result

/// The structured analysis result returned by Gemini.
struct AnalysisResult: Codable, Equatable {
    var keyPoints: [String]
    var reviewTable: [ReviewTableRow]
    var studyQuestions: [String]
    var flashcards: [Flashcard]
}

// MARK: - Review table row

/// A single row in the quick review table.
struct ReviewTableRow: Codable, Identifiable, Equatable {
    var id: UUID
    var concept: String
    var explanation: String

    init(id: UUID = UUID(), concept: String, explanation: String) {
        self.id = id
        self.concept = concept
        self.explanation = explanation
    }

    enum CodingKeys: String, CodingKey {
        case concept, explanation
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = UUID()
        self.concept = try container.decode(String.self, forKey: .concept)
        self.explanation = try container.decode(String.self, forKey: .explanation)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(concept, forKey: .concept)
        try container.encode(explanation, forKey: .explanation)
    }
}

// MARK: - Flashcard

/// A question/answer flashcard.
struct Flashcard: Codable, Identifiable, Equatable {
    var id: UUID
    var question: String
    var answer: String

    init(id: UUID = UUID(), question: String, answer: String) {
        self.id = id
        self.question = question
        self.answer = answer
    }

    enum CodingKeys: String, CodingKey {
        case question, answer
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = UUID()
        self.question = try container.decode(String.self, forKey: .question)
        self.answer = try container.decode(String.self, forKey: .answer)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(question, forKey: .question)
        try container.encode(answer, forKey: .answer)
    }
}

File: flutter_app/lib/models/analysis_result.dart
// MARK: - Analysis Result Models
// Mirrors Swift's AnalysisResult, ReviewTableRow, Flashcard structs
// from Sources/PDFAnalyzerAppCore/Models/AnalysisResult.swift

/// The structured analysis result returned by Gemini.
class AnalysisResult {
  final List<String> keyPoints;
  final List<ReviewTableRow> reviewTable;
  final List<String> studyQuestions;
  final List<Flashcard> flashcards;

  const AnalysisResult({
    required this.keyPoints,
    required this.reviewTable,
    required this.studyQuestions,
    required this.flashcards,
  });

  factory AnalysisResult.fromJson(Map<String, dynamic> json) {
    return AnalysisResult(
      keyPoints: List<String>.from(json['keyPoints'] as List? ?? []),
      reviewTable: ((json['reviewTable'] as List?) ?? [])
          .map((e) => ReviewTableRow.fromJson(e as Map<String, dynamic>))
          .toList(),
      studyQuestions: List<String>.from(json['studyQuestions'] as List? ?? []),
      flashcards: ((json['flashcards'] as List?) ?? [])
          .map((e) => Flashcard.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is AnalysisResult &&
          runtimeType == other.runtimeType &&
          keyPoints.toString() == other.keyPoints.toString() &&
          reviewTable.toString() == other.reviewTable.toString() &&
          studyQuestions.toString() == other.studyQuestions.toString() &&
          flashcards.toString() == other.flashcards.toString();

  @override
  int get hashCode =>
      keyPoints.hashCode ^ reviewTable.hashCode ^ studyQuestions.hashCode ^ flashcards.hashCode;
}

/// A single row in the quick review table.
class ReviewTableRow {
  final String id;
  final String concept;
  final String explanation;

  ReviewTableRow({
    String? id,
    required this.concept,
    required this.explanation,
  }) : id = id ?? _generateId();

  static int _idCounter = 0;

  static String _generateId() {
    _idCounter++;
    return 'row_${DateTime.now().millisecondsSinceEpoch}_$_idCounter';
  }

  factory ReviewTableRow.fromJson(Map<String, dynamic> json) {
    return ReviewTableRow(
      concept: json['concept'] as String? ?? '',
      explanation: json['explanation'] as String? ?? '',
    );
  }

  @override
  String toString() => 'ReviewTableRow(concept: $concept, explanation: $explanation)';
}

/// A question/answer flashcard.
class Flashcard {
  final String id;
  final String question;
  final String answer;

  Flashcard({
    String? id,
    required this.question,
    required this.answer,
  }) : id = id ?? _generateId();

  static int _idCounter = 0;

  static String _generateId() {
    _idCounter++;
    return 'card_${DateTime.now().millisecondsSinceEpoch}_$_idCounter';
  }

  factory Flashcard.fromJson(Map<String, dynamic> json) {
    return Flashcard(
      question: json['question'] as String? ?? '',
      answer: json['answer'] as String? ?? '',
    );
  }

  @override
  String toString() => 'Flashcard(question: $question, answer: $answer)';
}
